### ISSUE TITLE
"object-property-newline" rule: add option to allow shorthand properties on the same line

### ISSUE THREAD

**Issue by OliverSieweke** [eslint/eslint#11692]
**What rule do you want to change?**

[`object-property-newline`](https://eslint.org/docs/rules/object-property-newline)

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

A new `allowAllShorthandPropertiesOnSameLine` option:

- `allowAllShorthandPropertiesOnSameLine: true` allows all the shorthand properties to be on the same line.
- `allowAllShorthandPropertiesOnSameLine: false` (default) disallows shorthand properties to be on the same line.

[Currently the only possible option is `allowAllPropertiesOnSameLine`]

**Please provide some example code that this change will affect:**

```js
const object = {
    shortHandProperty1, shortHandProperty2, property3: "value3", property4: "value4"
};
```

**What does the rule currently do for this code?**

With `["error"]`, expects:

```js
const object = {
    shortHandProperty1,
    shortHandProperty2,
    property3: "value3",
    property4: "value4"
};
```

**What will the rule do after it's changed?**

With `["error", {allowAllShorthandPropertiesOnSameLine: true}]`, also validates:

```js
const object = {
    shortHandProperty1, shortHandProperty2,
    property3: "value3",
    property4: "value4"
};
```

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment**

I haven't found any style guide that explicitly recommends this, but it seems to be quite a common pattern to me.

**Comment by kaicataldo** [comment]
What happens if there are a lot of shorthand properties? It seems like you would be forced to do something like this (which I imagine could be viewed as undesirable, particular since it could conflict with `max-len`):

```js
const object = {
    shortHandProperty1, shortHandProperty2, shortHandProperty3, shortHandProperty4, shortHandProperty5, shortHandProperty6, shortHandProperty7, shortHandProperty8,
    property3: "value3",
    property4: "value4"
};
```

**Comment by OliverSieweke** [comment]
I don't think this would be a problem as the rule would _allow_ but not _enforce_ all shorthand properties to be on the same line, in exactly the same way as is currently done with the `allowAllPropertiesOnSameLine` option.

With `allowAllPropertiesOnSameLine` we currently we have:

```js
// Good!
// {"object-property-newline": ["error", {"allowAllPropertiesOnSameLine": true} ]}
const object = {
    property1: "value1", property2: "value2",
};
```
```js
// Bad! ["Line 4 exceeds the maximum line-length of 30. (max-len)"]
// {"object-property-newline": ["error", {"allowAllPropertiesOnSameLine": true} ], "max-len": ["error", {"code": 30}]}
const object = {
    property1: "value1", property2: "value2",
};
```
```js
// Good!
// {"object-property-newline": ["error", {"allowAllPropertiesOnSameLine": true} ], "max-len": ["error", {"code": 30}]}
const object = {
    property1: "value1", 
    property2: "value2",
};
```
With `allowAllShorthandPropertiesOnSameLine` we would likewise have:
```js
// Good!
// {"object-property-newline": ["error", {"allowAllShorthandPropertiesOnSameLine": true} ]}
const object = {
    shortHandProperty1, shortHandProperty2,
    property3: "value3",
    property4: "value4"
};
```
```js
// Bad! ["Line 4 exceeds the maximum line-length of 30. (max-len)"]
// {"object-property-newline": ["error", {"allowAllShorthandPropertiesOnSameLine": true} ], "max-len": ["error", {"code": 30}]}
const object = {
    shortHandProperty1, shortHandProperty2,
    property3: "value3",
    property4: "value4"
};
```
```js
// Good!
// {"object-property-newline": ["error", {"allowAllShorthandPropertiesOnSameLine": true} ], "max-len": ["error", {"code": 30}]}
const object = {
    shortHandProperty1, 
    shortHandProperty2,
    property3: "value3",
    property4: "value4"
};
```

**Comment by kaicataldo** [comment]
Thanks for explaining that - that make sense to me.

I have two more questions:

1) Can you illustrate, for clarity, the expected behavior with both the existing option and the new option on?

2) I wonder if changing the configuration option to `allowPropertiesOnSameLine: "all" | "shorthand"` would give us more flexibility in the future (for instance, if someone wanted to allow longhand properties on the same line we could very easily add a `"longhand"` option). It would also negate the complexity of using the two options simultaneously (though I could imagine someone wanting that). I’m not convinced that this suggestion is better, but want to at least discuss it so that we have a plan for this rule in the future.

**Comment by OliverSieweke** [comment]
## 1. 
I think that `allowAllShorthandPropertiesOnSameLine` and `allowAllShorthandPropertiesOnSameLine` can best be thought of as two independent options. They don't interact as there is no code that is simultaneously concerned by both options.

Independently of `allowAllShorthandPropertiesOnSameLine` the following would be valid with `allowAllShorthandPropertiesOnSameLine` set to `true` and invalid otherwise:
```js
const object = {
    shortHandProperty1,  shortHandProperty2,
    property3: "value3", 
    property4: "value4"
};
```
Likewise independently of `allowAllShorthandPropertiesOnSameLine` the following would be valid with 
`allowAllShorthandPropertiesOnSameLine` set to `true` and invalid otherwise:
```js
const object = {
    shortHandProperty1,  shortHandProperty2, property3: "value3", property4: "value4"
};
```
## 2. 
Conceptually it would make sense to group both options under a single `allowPropertiesOnSameLine` option. The way you describe it, too much flexibility would be lost for my taste. It would be nice to be able to use both options simultaneously (when I proposed the rule, that's what I had in mind). 

Maybe the option `allowPropertiesOnSameLine` could take an array as value with the possible elements being `"all"`, `"shorthand"`, `"longhand"`(?). E.g `"allowPropertiesOnSameLine": ["all", "shorthand"]`.
That way it would be quite straightforward for the user to reason about it: "Ok, I want to allow properties on the same line! Are all properties on the same line fine for me? (Yes/No), Are all shorthand properties on the same line fine for me? (Yes/No)..."
